File: frontend/src/api/ranks.ts
export interface RankRow {
  rank: number
  platform_player_id: string
  name: string
  position: string | null
  team: string | null
  adp: number | null
}

export interface RanksScope {
  season: string
  format: string
}

export interface RankSetSummary {
  id: number
  name: string
  platform: string
  season: string
  format: string
  player_count: number
}

export interface CreateRankSetParams {
  name: string
  season: string
  format: string
  seed_from_adp?: boolean
}

const API_BASE = import.meta.env.VITE_API_BASE ?? 'http://127.0.0.1:8000'

async function parseOrThrow<T>(response: Response, label: string): Promise<T> {
  if (!response.ok) {
    let detail = ''
    try {
      const body = (await response.json()) as { detail?: string }
      detail = body.detail ?? ''
    } catch {
      // response body wasn't JSON -- fall back to just the status code
    }
    throw new Error(
      detail ? `${label}: ${detail}` : `${label} failed: ${response.status}`,
    )
  }
  return response.json() as Promise<T>
}

/** ADP-fallback read used by the draft player pool -- see the backend's
 * resolve_rank_set() for the "which rank set does this mean" caveat. */
export async function fetchRanks(scope: RanksScope): Promise<RankRow[]> {
  const query = new URLSearchParams({
    season: scope.season,
    format: scope.format,
  })
  const response = await fetch(`${API_BASE}/ranks?${query.toString()}`)
  return parseOrThrow(response, 'Fetching ranks')
}

export async function fetchRankSets(
  scope: RanksScope,
): Promise<RankSetSummary[]> {
  const query = new URLSearchParams({
    season: scope.season,
    format: scope.format,
  })
  const response = await fetch(`${API_BASE}/rank-sets?${query.toString()}`)
  return parseOrThrow(response, 'Fetching rank sets')
}

export async function createRankSet(
  params: CreateRankSetParams,
): Promise<RankSetSummary> {
  const response = await fetch(`${API_BASE}/rank-sets`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(params),
  })
  return parseOrThrow(response, 'Creating rank set')
}

export async function renameRankSet(
  rankSetId: number,
  name: string,
): Promise<RankSetSummary> {
  const response = await fetch(`${API_BASE}/rank-sets/${rankSetId}`, {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
  return parseOrThrow(response, 'Renaming rank set')
}

export async function deleteRankSet(rankSetId: number): Promise<void> {
  const response = await fetch(`${API_BASE}/rank-sets/${rankSetId}`, {
    method: 'DELETE',
  })
  if (!response.ok) {
    throw new Error(`Deleting rank set failed: ${response.status}`)
  }
}

export async function fetchRanksForSet(rankSetId: number): Promise<RankRow[]> {
  const response = await fetch(`${API_BASE}/rank-sets/${rankSetId}/ranks`)
  return parseOrThrow(response, 'Fetching ranks')
}

export async function saveRanksForSet(
  rankSetId: number,
  platformPlayerIds: string[],
): Promise<{ count: number }> {
  const response = await fetch(`${API_BASE}/rank-sets/${rankSetId}/ranks`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ platform_player_ids: platformPlayerIds }),
  })
  return parseOrThrow(response, 'Saving ranks')
}

File: backend/app/api/ranks.py
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Response
from pydantic import BaseModel
from sqlalchemy.orm import Session

from app.db import get_db
from app.ranks import (
    RankSetError,
    create_rank_set,
    delete_rank_set,
    list_rank_sets,
    list_ranks,
    rename_rank_set,
    replace_ranks,
    resolve_rank_set,
)

router = APIRouter()

DbSession = Annotated[Session, Depends(get_db)]

DEFAULT_SEASON = "2026"
DEFAULT_FORMAT = "half_ppr"


class RankSetSummary(BaseModel):
    id: int
    name: str
    platform: str
    season: str
    format: str
    player_count: int


class CreateRankSetRequest(BaseModel):
    name: str
    season: str = DEFAULT_SEASON
    format: str = DEFAULT_FORMAT
    platform: str = "sleeper"
    seed_from_adp: bool = True


class RenameRankSetRequest(BaseModel):
    name: str


class RankRow(BaseModel):
    rank: int
    platform_player_id: str
    name: str
    position: str | None
    team: str | None
    adp: float | None


class ReplaceRanksRequest(BaseModel):
    platform_player_ids: list[str]


class ReplaceRanksResponse(BaseModel):
    count: int


@router.get("/rank-sets", response_model=list[RankSetSummary])
def get_rank_sets(
    db: DbSession,
    platform: str = "sleeper",
    season: str | None = None,
    format: str | None = None,
) -> list[RankSetSummary]:
    rows = list_rank_sets(db, platform, season, format)
    return [RankSetSummary(**row) for row in rows]


@router.post("/rank-sets", response_model=RankSetSummary)
def post_rank_set(payload: CreateRankSetRequest, db: DbSession) -> RankSetSummary:
    try:
        rank_set = create_rank_set(
            db,
            payload.name,
            payload.season,
            payload.format,
            platform=payload.platform,
            seed_from_adp=payload.seed_from_adp,
        )
    except RankSetError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    rows = list_rank_sets(db, payload.platform, payload.season, payload.format)
    summary = next(row for row in rows if row["id"] == rank_set.id)
    return RankSetSummary(**summary)


@router.patch("/rank-sets/{rank_set_id}", response_model=RankSetSummary)
def patch_rank_set(
    rank_set_id: int, payload: RenameRankSetRequest, db: DbSession
) -> RankSetSummary:
    try:
        rank_set = rename_rank_set(db, rank_set_id, payload.name)
    except RankSetError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    rows = list_rank_sets(db, rank_set.platform, rank_set.season, rank_set.format)
    summary = next(row for row in rows if row["id"] == rank_set.id)
    return RankSetSummary(**summary)


@router.delete("/rank-sets/{rank_set_id}", status_code=204)
def delete_rank_set_route(rank_set_id: int, db: DbSession) -> Response:
    try:
        delete_rank_set(db, rank_set_id)
    except RankSetError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return Response(status_code=204)


@router.get("/rank-sets/{rank_set_id}/ranks", response_model=list[RankRow])
def get_rank_set_ranks(rank_set_id: int, db: DbSession) -> list[RankRow]:
    rows = list_ranks(db, rank_set_id)
    return [RankRow(**row) for row in rows]


@router.put("/rank-sets/{rank_set_id}/ranks", response_model=ReplaceRanksResponse)
def put_rank_set_ranks(
    rank_set_id: int, payload: ReplaceRanksRequest, db: DbSession
) -> ReplaceRanksResponse:
    count = replace_ranks(db, rank_set_id, payload.platform_player_ids)
    return ReplaceRanksResponse(count=count)


@router.get("/ranks", response_model=list[RankRow])
def get_ranks(
    db: DbSession,
    platform: str = "sleeper",
    season: str = DEFAULT_SEASON,
    format: str = DEFAULT_FORMAT,
) -> list[RankRow]:
    """Resolver-backed read used by the draft player pool's ADP-fallback fetch --
    "the ranks for this format" (see resolve_rank_set's docstring for the caveat).
    """
    rank_set = resolve_rank_set(db, platform, season, format)
    if rank_set is None:
        return []
    rows = list_ranks(db, rank_set.id)
    return [RankRow(**row) for row in rows]
